ProfitApp/src/main/java/bsu/fpmi/profit/ProfitCRUD.java

```diff
@@ -0,0 +1,424 @@
+        preparedStatement.setInt(7, Integer.parseInt(offer.getDiscount()));
+        preparedStatement.setTimestamp(8, new java.sql.Timestamp(offer.getCreatedAt().getTime()));
+        preparedStatement.setString(9, offer.getPhotoLink());
+        return preparedStatement;
+    }
+
+    private void addHashtags(Ad offer, int offerId) throws SQLException {
+        String sql = "INSERT INTO HASHTAGS(OFFER_ID, TAG) VALUES (?, ?);";
+        for (String tag : offer.getHashTags()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(sql);
+            preparedStatement.setInt(1, offerId);
+            preparedStatement.setString(2, tag);
+            preparedStatement.executeUpdate();
+            preparedStatement.close();
+        }
+    }
+
+    public boolean addOffer(Ad offer){
+        if(validateOffer(offer)) {
+            try {
+
+                String sql = "select user_id from user where username = ?";
+                PreparedStatement preparedStatement = connection.prepareStatement(sql);
+                preparedStatement.setString(1, offer.getVendor());
+                ResultSet resultSet = preparedStatement.executeQuery();
+
+
+                int userId;
+                if (resultSet.next()) {
+                    userId = resultSet.getInt("user_id");
+                    preparedStatement.close();
+                    int offerId = (int) (offer.getCreatedAt().getTime() % Integer.MAX_VALUE);
+
+                    preparedStatement = buildAddPreparedStatement(offerId, userId, offer);
+
+                    preparedStatement.executeUpdate();
+                    preparedStatement.close();
+
+                    addHashtags(offer, userId);
+
+                    return true;
+                }
+            } catch (SQLException e) {
+                e.printStackTrace();
+            }
+        }
+        return false;
+    }
+
+    private String getEditSqlString(Ad ad){
+        StringBuilder sb = new StringBuilder("update offer set ");
+        if(ad.getLabel() != null && ad.getLabel().length() != 0){
+            sb.append("OFFER_NAME = ?, ");
+        }
+        if (ad.getDescription() != null && ad.getDescription().length() != 0) {
+            sb.append("DESCRIPTION = ?, ");
+        }
+        if (ad.getLink() != null) {
+            sb.append("VENDOR_LINK = ?, ");
+        }
+        if (ad.getPhotoLink() != null) {
+            sb.append("PHOTO_LINK = ?, ");
+        }
+        if (ad.getDiscount() != null) {
+            sb.append("DISCOUNT = ?, ");
+        }
+        if (ad.getValidUntil() != null) {
+            sb.append("VALID_UNTIL = ?, ");
+        }
+        sb.deleteCharAt(sb.length() - 2);
+        sb.append("where offer_id = ?");
+        return sb.toString();
+    }
+
+    private void editHashtags(String id, Ad ad) throws SQLException{
+        String sql = "delete from hashtags where offer_id = ?";
+        PreparedStatement preparedStatement = connection.prepareStatement(sql);
```

Comment:
Use "try-with-resources" struct
```
try (PreparedStatement statement = ...) {
   ...
} catch (Exception ex) {
  ... 
}
```
It will be properly closed regardless of an exception